#!/bin/bash

#PBS -l mem=8gb,nodes=1:ppn=8,walltime=16:00:00
#PBS -m abe
#PBS -M user@example.com
#PBS -q lab

set -e
set -u
set -o pipefail

module load parallel

#   This script is a QSub submission script for converting a batch of SAM files to BAM files
#   To use, on line 5, change the 'user@example.com' to your own email address
#       to get notificaitons on start and completion of this script
#   Define a path to SAMTools in the 'SAMTOOLS' field on line 38
#       If using MSI, leave the definition as is to use their installation of SAMTools
#       Otherwise, it should look like this:
#           SAMTOOLS=${HOME}/software/samtools
#       Please be sure to comment out (put a '#' symbol in front of) the 'module load samtools' on line 37
#       And to uncomment (remove the '#' symbol) from the 'SAMTOOLS=' on line 38
#   Add the full file path to list of samples on the 'SAMPLE_INFO' field on line 41
#       This should look like:
#           SAMPLE_INFO=${HOME}/Directory/list.txt
#       Use ${HOME}, as it is a link that the shell understands as your home directory
#           and the rest is the full path to the actual list of samples
#   Define a path to a reference genome on line 47
#       This should look like:
   #        REF_GEN=${HOME}/Directory/reference_genome.fa
#   Put the full directory path for the output in the 'SCRATCH' field on line 50
#       This should look like:
#           SCRATCH="${HOME}/Out_Directory"
#       Adjust for your own out directory.
#   Name the project in the 'PROJECT' field on line 53
#       This should look lke:
#           PROJECT=Genetics

#   Load the SAMTools module for MSI, else define path to SAMTools installation
module load samtools
#SAMTOOLS=

#   List of SAM files for conversion
SAMPLE_INFO=

#   Reference genome to help base the conversion off of
REF_GEN=

#   Scratch directory, for output
SCRATCH=

#   Name of project
PROJECT=

#   Make the outdirectories
OUT=${SCRATCH}/${PROJECT}
mkdir -p ${OUT}/stats ${OUT}/deduped ${OUT}/sorted ${OUT}/finished ${OUT}/raw

#   Generate a list of sample names
for i in `seq $(wc -l < "${SAMPLE_INFO}")`
do
    s=`head -"$i" "${SAMPLE_INFO}" | tail -1`
    basename "$s" .sam >> "${SCRATCH}"/"${PROJECT}"/sample_names.txt
done

SAMPLE_NAMES="${SCRATCH}"/"${PROJECT}"/sample_names.txt

#   Define a function to process SAM files generated by BWA mem
#       This will create the following files for each input SAM file:
#           a sorted BAM file
#           alignment statistics for the sorted BAM file
#           a deduplicated BAM file
#           alignment statistics for the deduplicated BAM file
function process_sam() {
    module load samtools
    #   Today's date
    YMD=`date +%Y-%m-%d`
    SAMFILE="$1"
    REF_SEQ="$2"
    OUTDIR="$3"
    #   Sample name, taken from full name of SAM file
    SAMPLE_NAME=`basename "${SAMFILE}" .sam`
    #   Generate a sorted BAM file
    samtools view -bT "${REF_GEN}" "${SAMFILE}" > "${OUTDIR}"/raw/"${SAMPLE_NAME}"_"${YMD}"_raw.bam
    #   Create alignment statistics for the raw BAM file
    samtools flagstat "${OUTDIR}"/raw/"${SAMPLE_NAME}"_"${YMD}"_raw.bam > "${OUTDIR}"/stats/"${SAMPLE_NAME}"_"${YMD}"_raw_stats.out
    #   Sort the raw BAM file
    samtools sort "${OUTDIR}"/raw/"${SAMPLE_NAME}"_"${YMD}"_raw.bam "${OUTDIR}"/sorted/"${SAMPLE_NAME}"_"${YMD}"_sorted.bam
    #   Deduplicate the sorted BAM file
    samtools rmdup "${OUTDIR}"/sorted/"${SAMPLE_NAME}"_"${YMD}"_sorted.bam "${OUTDIR}"/finished/"${SAMPLE_NAME}"_"${YMD}"_finished.bam
    #   Create alignment statistics using SAMTools
    samtools flagstat "${OUTDIR}"/finished/"${SAMPLE_NAME}"_"${YMD}"_finished.bam > "${OUTDIR}"/stats/"${SAMPLE_NAME}"_"${YMD}"_finished_stats.out
}

#   Export the SAM file processing function to be used by GNU Parallel
export -f process_sam

#   Run the SAM file processing function in parallel for all input SAM files
cat ${SAMPLE_INFO} | parallel "process_sam {} ${REF_GEN} ${OUT}"

#   Create a sorted BAM file for each input SAM file in parallel
#parallel --xapply "samtools view -bT ${REF_GEN} {1} | samtools sort - - | tee >(samtools flagstat - > ${SCRATCH}/${PROJECT}/stats/{2}.out) > ${SCRATCH}/${PROJECT}/{2}_${DATE}.bam" :::: ${SAMPLE_INFO} :::: ${SAMPLE_NAMES}

#   Make a list of deduplicated BAM files for further use
find ${SCRATCH}/${PROJECT}/deduped -name "*.bam" | sort > ${SCRATCH}/${PROJECT}/${PROJECT}_bam_files.txt
echo "List of BAM files can be found at"
echo "${SCRATCH}/${PROJECT}/${PROJECT}_bam_files.txt"
